Wizard V2: Custom Repositories refactor

Andrewgdewar opened the pull request:
- Updated content api usage
- Refactored repositories
- A lot of small changes 

One design change: 
My pm suggested removing the confirmation modal on deletion during the edit screen, I have commented that out and will ask for feedback from all of you. We believe the warning below should be sufficient.
<img width="619" alt="Screenshot 2024-05-06 at 10 46 51 PM" src="https://github.com/osbuild/image-builder-frontend/assets/38083295/d77d7239-59a1-40c6-9be9-6847e7c2f8ed">

Also updated the empty state screen: 
<img width="1110" alt="Screenshot 2024-05-08 at 8 34 19 PM" src="https://github.com/osbuild/image-builder-frontend/assets/38083295/6a81463c-a3c0-45e8-aa3c-9b7b2d3e4b31">

<img width="1120" alt="Screenshot 2024-05-08 at 8 39 01 PM" src="https://github.com/osbuild/image-builder-frontend/assets/38083295/0d9e564a-9852-4265-87ea-e8db35304f92">

regexowl commented on src/test/Components/CreateImageWizardV2/CreateImageWizard.content.test.tsx:
```diff
@@ -475,36 +475,9 @@ describe('Step Custom repositories', () => {
+    await screen.findByText(/select page \(10 items\)/i);
   });
 
-  test('filter works', async () => {
```
I don't think we should be removing tests that don't pass even if it's after a big refactor. We'll still want to test if filtering repositories works.

When manually testing the filtering works well on real life data so in this case it's the MSW handlers/fixtures that will need to be updated for the mocked values to pass the tests as well. Since this might be a bigger task it should be okay to comment out the failing tests and split the fixtures updates to a separate PR

regexowl commented on src/test/Components/CreateImageWizardV2/CreateImageWizard.content.test.tsx:
```diff
@@ -475,36 +475,9 @@ describe('Step Custom repositories', () => {
+    await screen.findByText(/select page \(10 items\)/i);
   });
 
-  test('filter works', async () => {
```
I don't think we should be removing tests that don't pass even if it's after a big refactor. We'll still want to test if filtering repositories works.

When manually testing the filtering works well on real life data so in this case it's the MSW handlers/fixtures that will need to be updated for the mocked values to pass the tests as well. Since this might be a bigger task it should be okay to comment out the failing tests and split the fixtures updates to a separate PR

regexowl replied:
Aah, yeah, you're absolutely right. I didn't realize that this would now test API responses instead of filtering within the frontend. Mea culpa :sweat_smile:

regexowl commented on src/test/Components/CreateImageWizardV2/CreateImageWizard.content.test.tsx:
```diff
@@ -475,36 +475,9 @@ describe('Step Custom repositories', () => {
     );
 
-    await screen.findByText(/select all \(1016 items\)/i);
+    await screen.findByText(/select page \(10 items\)/i);
```
I've discussed with @lucasgarfield previously that testing on 1000 of mocked repositories might be a bit too much. Maybe we could also update the fixture to generate less filler repos? Just a though, might not be completely in the scope of this PR :sweat_smile:

regexowl replied:
Agree, let's stick with 10 here. The fixtures can be updated later if needed.

regexowl commented on src/test/Components/CreateImageWizardV2/steps/Repositories/Repositories.test.tsx:
```diff
@@ -102,28 +94,6 @@ describe('repositories request generated correctly', () => {
     },
   ];
 
-  const expectedNginxRepository: Repository = {
```
Same here, as far as I'm aware this might be the only test with module_hotfixes. For the sake of not loosing coverage the MSW fixtures should be updated to accommodate the changes to the code and reflect how real data works.

Could be probably split into a separate PR, but the test should be commented out with a note rather than just removed.

Andrewgdewar replied:
Fixed, I mocked searching.

regexowl commented on src/Components/CreateImageWizardV2/steps/Repositories/Repositories.tsx:
```diff
@@ -40,224 +43,173 @@ import {
       contentType: 'rpm',
       origin: 'external',
-      limit: 100,
+      limit: perPage,
```
Previous double fetching method is not very ideal, but I wonder if we want to fetch based on perPage limit only as any changes cause more loading states. Will return to this in the main comment.

The pagination doesn't seem to work:
![broken_pagination](https://github.com/osbuild/image-builder-frontend/assets/49452678/0954a8e7-826a-42c2-981a-cf73af3d16e3)

ezr-ondrej replied:
Probably not caused by this PR: https://github.com/osbuild/image-builder-frontend/issues/2084 :)

regexowl replied:
I believe those are two separate issues as currently the pagination on the Repositories step seems to work. The proposed changes in this PR would break it.

ezr-ondrej replied:
Ah sorry! I've gotten confused, you're right, those are separate issues!

regexowl replied:
No worries :)

Andrewgdewar replied:
> Ah sorry! I've gotten confused, you're right, those are separate issues!

Actually same issue, I keep forgetting both the page/perPage are needed references for useMemo/dependency arrays.

regexowl reviewed (commented):
Thank you for working on the refactor! The code looks much cleaner :heart: 

Looking at this from the user point of view there seems to be considerably more loading as more actions trigger a new API call including search and removing repositories.

This is linked to the question of how many repositories should we initially fetch - previously we fetched all the repositories right from the start which looked a bit more seamless when going through the step. What I mean is this:

- previously the fetching happened at the initialization of the step, but then there was no further loading / skeletons:
![repos-old](https://github.com/osbuild/image-builder-frontend/assets/49452678/ba2204ef-bd11-4a94-8176-41256fc44b16)

- now the loading state gets triggered on toggling between all/selected, searching, clearing search input, changing pagination perPage limit and removing a repo
![repos-new](https://github.com/osbuild/image-builder-frontend/assets/49452678/f7d1785b-4bd1-4982-b8ee-cd0728780342)

Which might not be necessarily bad, but it might be good to discuss which path to take from this point. What would be the advantages of handling the data via a series of API calls?

regexowl commented:
As for the feedback to removing the modal - I remember adding it for some reason, but can't really remember what the reason was :sweat_smile:

I guess it was there to make double sure that the user knows they can break their image with the change? Not sure if that's too much safeguarding though.

Any thoughts @mmenestr @kelseamann ?

Also to check - when removing a repository what would be the desired behaviour? Currently when in the Selected toggle the checkbox gets unchecked, but the repository stays there until the toggle gets toggled to All and retoggled to Selected again.

After the changes the repository would disappear from the Selected list as soon as it gets unchecked. I'm aware we're not currently consistent with the behaviour between repos / packages steps, so this might be an issue for a separate PR.

mmenestr commented:
I feel like we should keep the modal... this is probably a rare scenario anyway. Also, I think that when the user unselects the repo, the packages should automatically get removed from the "selected" toggle.

In general - my guidance when it comes to unselected items is this:

If the user is in the "selected" toggle, and unchecks something, then the items should remain in the "selected" list, unchecked, until the user toggles back to "All".

However, if the user is NOT on the selected toggle, and their actions impact the selected toggle, then the items should automatically go away from the selected toggle. 

Does that make sense? It basically just depends on whether the user is in the selected toggle while things are unselected or not.

regexowl commented:
> Also, I think that when the user unselects the repo, the packages should automatically get removed from the "selected" toggle.

Agree, alas that's currently not possible as the packages aren't linked to their origin repository.

Andrewgdewar commented:
> As for the feedback to removing the modal - I remember adding it for some reason, but can't really remember what the reason was 😅
> 
> I guess it was there to make double sure that the user knows they can break their image with the change? Not sure if that's too much safeguarding though.

> 
> Any thoughts @mmenestr @kelseamann ?
> 
> edit: it should be this modal if I'm correct ![image](https://private-user-images.githubusercontent.com/49452678/332766800-95abb4f1-0ba2-40a1-8160-9063a16dfe5e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5MDgxNTYsIm5iZiI6MTcxNjkwNzg1NiwicGF0aCI6Ii80OTQ1MjY3OC8zMzI3NjY4MDAtOTVhYmI0ZjEtMGJhMi00MGExLTgxNjAtOTA2M2ExNmRmZTVlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI4VDE0NTA1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM5MGNlODNlMjNiZTNiOTVmYzc3ZThlYTNlOTRiNTNhMDI2YmM3N2ViZTMxYmQxMTdkZTJiYzdkNWE4ZTk3NjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Nyf9wE-oWsKRaCtFSkqzR-TFskMr3nIT5PqrmqDwJtg)
> 

I spoke to @jlsherrill regarding this (SME for content) and he had basically said that the warning should be enough. 
But if you would like this to be re-added I can do!


> Also to check - when removing a repository what would be the desired behaviour? Currently when in the Selected toggle the checkbox gets unchecked, but the repository stays there until the toggle gets toggled to All and retoggled to Selected again.
> 
> After the changes the repository would disappear from the Selected list as soon as it gets unchecked. I'm aware we're not currently consistent with the behaviour between repos / packages steps, so this might be an issue for a separate PR.


Oh hmm, I see what you mean. 
That is the exact behaviour we have today in content templates when adding and removing a repository for reference. 
I may be able to add a debounce on the selected items to alleviate that extra loading.

regexowl commented:
> The primary reason I am using the API in this way is consistency across our apps. As it was the API was built around this functionality.

Got it, sounds great!

> I understand the idea of liking to front-load all data, but in general, it is largely frowned upon as an architectural choice unless there is an inherit slowness with the API.
> 
> The usual reasons this is sited as not a good pattern in general is you can get to a point where with enough data it becomes a problem, there are also different data limits across browsers and devices, that can lock up/slow-down the web page dramatically (likely not an issue but I digress).

Fair enough! I don't know how many repositories users usually have, but I admit that relying on the idea that "there can't be that many" is not a good approach. :sweat_smile: 

We can try to figure out if there's a way to cache some data so the loading is less obvious, but I'd say that's out of the scope of this PR for now.

Thanks a lot for the explanation!

regexowl commented:
> But with the addition of the remove modal, it may be less obvious as the user will be slowed down further.

Hmm, when you're looking at the debounce would it be possible to make a short screencast with and without the modal?

I think there are two open questions regarding the modal:
- how to make it obvious to the user that removing previously used repositories is a really bad idea
- and in case the way is to display the modal, when should it be displayed?

Since we probably want the newly unchecked repos to stay in Selected for a bit, it may indeed clash with the modal popping out as soon as the repo gets unselected. 

Maybe we could show the modal just once when the Next button is clicked after the changes were applied? This would probably need to be handled via wizard footer then... :thinking: 

Or maybe we could display the alert only when a previously used repo was unselected so it pops out more?

What do you think @mmenestr?

regexowl commented:
Clicked around a bunch - this looks and works great, thank you! Current way of fetching/refreshing feels much better to me! The skeletons were a bit overwhelming :skull: 

I'd say let's double check the module_hotfixes test, decide on the modal and debounce and then it's good to get merged.

We'll need to update msw fixtures and handlers a bit if we want to keep making use of the in-browser mocking, but that's a separate problem. Opened an issue for that: https://github.com/osbuild/image-builder-frontend/issues/2107

regexowl commented:
Oh, one thing - could we make the search case insensitive?

![image](https://github.com/osbuild/image-builder-frontend/assets/49452678/9e26ca27-f364-4ae2-971f-685419d3e314)

![image](https://github.com/osbuild/image-builder-frontend/assets/49452678/ac8462dc-5c39-43cd-8e00-6f853ef30f86)

Andrewgdewar commented:
> Oh, one thing - could we make the search case insensitive?
> 
> ![image](https://private-user-images.githubusercontent.com/49452678/334763322-9e26ca27-f364-4ae2-971f-685419d3e314.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5ODA0ODksIm5iZiI6MTcxNjk4MDE4OSwicGF0aCI6Ii80OTQ1MjY3OC8zMzQ3NjMzMjItOWUyNmNhMjctZjM2NC00YWUyLTk3MWYtNjg1NDE5ZDNlMzE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI5VDEwNTYyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJmOGI5YWI3OGJkNzQ4OTkyMmE5MGM4NGZmZmY3ZjM0Mzk0YTkwNzZlMGZkNDQwZDc5ZDM1MjZmZDY3ZTI1ZTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.8vA4jn5UmJmHUK3oz5Ltk0ODOgp6oVvMG3qWahEAgL8)
> 
> ![image](https://private-user-images.githubusercontent.com/49452678/334763415-ac8462dc-5c39-43cd-8e00-6f853ef30f86.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5ODA0ODksIm5iZiI6MTcxNjk4MDE4OSwicGF0aCI6Ii80OTQ1MjY3OC8zMzQ3NjM0MTUtYWM4NDYyZGMtNWMzOS00M2NkLThlMDAtNmY4NTNlZjMwZjg2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI5VDEwNTYyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwOGI2MjdjNmEzNGUxMGU3NDM4NmZiNzJhZjJkYzVmZjA4OTQzZDNjMDgzZGZiZTE0YTFhZjQyZDhjNTBmNDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.WEU6tedtD7yOyxaOK9ho5qHU3wIDrn2MBpWY2WmOSGo)

Good sanity check. 
Added a ticket for the API [here](https://issues.redhat.com/browse/HMS-4222).

ezr-ondrej commented:
/retest

Andrewgdewar commented:
Added the modal, made it so I only warn if on initial state. 
Also changed the wording a bit as with the warning in the background, it felt a bit redundant.
![Screenshot 2024-05-29 at 3 00 22 PM](https://github.com/osbuild/image-builder-frontend/assets/38083295/b03e2222-28ec-47c1-ba8d-a8b213a6ac5c)

I also fixed some loading/state issues with the packages tab.

regexowl commented:
/retest

regexowl commented:
/retest

regexowl commented on src/Components/CreateImageWizardV2/steps/Repositories/Repositories.tsx:
```diff
@@ -299,418 +337,221 @@ const Repositories = () => {
         ]}
       >
-        We do not recommend removing any repositories if you have added packages
+        You are removing a previously added repository
```
A teeny tiny nitpick.

```suggestion
        You are removing a previously added repository.
```

regexowl commented on src/Components/CreateImageWizardV2/steps/Repositories/components/BulkSelect.tsx:
```diff
@@ -0,0 +1,86 @@
+  isDisabled: boolean;
+}
+
+export function BulkSelect({
```
When removing the page of repos via bulk select the remove modal gets triggered even for new blueprint

![repos-modal](https://github.com/osbuild/image-builder-frontend/assets/49452678/ea849ed6-a1a5-4a1c-9712-0987f7021aef)

Andrewgdewar replied:
Hmm will fix this.

regexowl reviewed (commented):
Found one bug and a tiny nitpick, but otherwise looks great to me!

Andrewgdewar commented:
Okay, so I fixed the bug above, and looked into cacheing [here](https://redux-toolkit.js.org/rtk-query/usage/cache-behavior). 
I then rewrote a few things and the cache/refetch functions now work a LOT better. 
I think the snappiness you are offer is largely back! 🎉

Andrewgdewar commented:
I have added the "search" field to the repositories mock function to get around the test failures.

Andrewgdewar commented:
Fixed lint and flattened merge history.

regexowl commented:
/retest

regexowl commented:
/retest

regexowl reviewed (approved):
Looks awesome and works like magic! :mage: Thanks for all your work on this!!
